### PR TITLE
Improve keygen grind algo by 10%

### PR DESCRIPTION
**Pull request by jarry-xiao**
#### Problem
`solana-keygen grind` is slightly inefficient in that it checks a few too many keys when searching for prefixes. The largest pubkey is `JEKNVnkbo3jma5nREBBJCDoXFVeKkD56V3xKrvRmWxFG` and has 44 characters. If the starting ascii character for the grind is between K and z, this means that it is impossible for the resulting pubkey to match if the sampled key has 44 characters.

The smallest key with 44 characters is `21111111111111111111111111111111111111111111`, we can short circuit the main loop if the sampled key >= `21111111111111111111111111111111111111111111` depending on the supplied prefixes.

#### Summary of Changes
Efficiency improvement is admittedly marginal:

Original:
```
Searched 40000000 keypairs in 93s. 0 matches found.
```

Updated:
```
Searched 40000000 keypairs in 81s. 0 matches found.
```

I guess bs58 encode is faster than i expected and most of the bottleneck is probably in `ed25519_dalek::Keypair::generate`. Still a reasonably fun/quick 30 min project 

Fixes #

### CODE REVIEW

**Comment by diman-io** [comment]
JFI

https://github.com/solana-labs/solana/pull/29786

I made PR and forgot :) But it turned out that CI couldn't :(

@jarry-xiao Could you please check, may be they might be joined?
The main idea in that PR is to compare bytes before base58 decoding.

File: keygen/src/keygen.rs [hunk 1]
```diff
@@ -769,6 +796,10 @@ fn do_main(matches: &ArgMatches) -> Result<(), Box<dyn error::Error>> {
                         } else {
                             (Keypair::new(), "".to_string())
                         };
+                        // Skip keypairs that will never match the user specified prefix
```
Comment (mvines) [inline keygen/src/keygen.rs:799]:
```suggestion
                        // Skip keypairs that will never match the user specified prefix
```

**Review by mvines — DISMISSED** [review]

**Comment by mvines** [comment]
Boo, my review doesn't count anymore since I fixed up the whitespace in the original commit so I guess I'm now an author. @t-nelson, can you help a brother out with an approval here?

File: keygen/src/keygen.rs [hunk 2]
```diff
@@ -37,6 +37,18 @@ use {
     },
 };
 
+mod smallest_length_44_public_key {
+    use solana_sdk::declare_id;
+
+    declare_id!("21111111111111111111111111111111111111111111");
```
Comment (t-nelson) [inline keygen/src/keygen.rs:43]:
```suggestion
    pubkey!("21111111111111111111111111111111111111111111");
```

more accurate as we're not talking about a program ID here

File: keygen/src/keygen.rs [hunk 3]
```diff
@@ -729,6 +741,20 @@ fn do_main(matches: &ArgMatches) -> Result<(), Box<dyn error::Error>> {
             };
             let no_outfile = matches.is_present(NO_OUTFILE_ARG.name);
 
+            // The vast majority of base58 encoded public keys have length 44, but
+            // these only enacapsulate prefixes 1-9 and A-H.  If the user is searching
+            // for a keypair that starts with a prefix of J-Z or a-z, then there is no
+            // reason to waste time searching for a keypair that will never match
+            let skip_len_44_pubkeys = grind_matches
```
Comment (t-nelson) [inline keygen/src/keygen.rs:748]: it'd probably pay to warn the user that they're searching for an unlikely prefix

Comment (jarry-xiao) [reply keygen/src/keygen.rs:748]: Majority of prefixes are pretty unlikely :)

Comment (diman-io) [reply keygen/src/keygen.rs:748]: Looks like this does not take into account the `--ignore-case` argument

File: keygen/src/keygen.rs [hunk 4]
```diff
@@ -729,6 +741,20 @@ fn do_main(matches: &ArgMatches) -> Result<(), Box<dyn error::Error>> {
...
+                .iter()
+                .map(|g| {
+                    let target_key = g.starts.clone()
```
Comment (t-nelson) [inline keygen/src/keygen.rs:751]:
```suggestion
                    let target_key = if ignore_case {
                        g.starts.to_ascii_uppercase()
                    } else {
                        g.starts.clone()
                    };
```

to handle `--ignore-case`

File: keygen/src/keygen.rs (hunk 3 above)
Comment (t-nelson) [reply keygen/src/keygen.rs:748]: > Looks like this does not take into account the `--ignore-case` argument

https://github.com/solana-labs/solana/pull/30365/files#r1158852602

File: keygen/src/keygen.rs [hunk 5]
```diff
@@ -729,6 +741,20 @@ fn do_main(matches: &ArgMatches) -> Result<(), Box<dyn error::Error>> {
             };
             let no_outfile = matches.is_present(NO_OUTFILE_ARG.name);
 
+            // The vast majority of base58 encoded public keys have length 44, but
+            // these only enacapsulate prefixes 1-9 and A-H.  If the user is searching
```
Comment (kevinji) [inline keygen/src/keygen.rs:745]:
```suggestion
            // these only encapsulate prefixes 1-9 and A-H.  If the user is searching
```

**Review by t-nelson — APPROVED** [review]

**Pull request merged**